### ISSUE TITLE
Performance improvements for URL patterns

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#4409]
I've had various ideas for performance improvements for [URL patterns](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/utils/urlmatch.py):

# UrlPatternSet

From [this mail](https://lists.schokokeks.org/pipermail/qutebrowser/2018-November/000533.html):

The best way out would be having some kind of `UrlPatternSet` class which can be more intelligent about lookups. I've had some thoughts on that noted down somewhere already, but I can't find them right now. Basically, you could hash patterns by their hosts, and then if you want to check whether `foobar.example.com` is contained, make some O(1) lookups for:

- `foobar.example.com`
- `*.example.com`
- `*.com`
- `*`

However, ensuring that this works efficiently for all imaginable cases probably requires some more thought. Note that we'd need to take care of keeping the rule order somehow (so we know which rule has priority).

I'd instead propose the following:

- Add a `UrlPatternSet` class which we can extend like explained above over time.
- For now, just add a special case for patterns in the form of `*://example.com/*` (all schemes, constant host, all paths), by looking them up in a Python set. For all patterns which are more complex than that, fall back to iterating.

That solution should be quite straightforward for now, and result in a huge speedup for the common case of adblocking lists. 

# Speed up creating pattern

We should profile creating `UrlPattern` objects and see what we can speed up. IIRC I did so a while ago, and `if not utils.raises(ValueError, ipaddress.ip_address, host):` took quite some time. Maybe we can first add a heuristic like checking whether the there's a `:` in the URL or the first char is a digit?

# Speed up path matching

Currently we use `fnmatch` which uses regexes under the hood. Chromium seems to have a more efficient implementation.

**Comment by jgkamat** [comment]
Lots of discussion on IRC so I'm dumping it all in one place:

```
[21:37:47] <jgkamat> The-Compiler: fwiw, I made a simple example here. It looks like re in python is really, really slow, but the 'regex' package in pypi is over an order of magnitude faster than iteration. I'm wondering if we can use a real regexp matching library in python (like re2), which should be even faster. In theory, the tree matching and the regexp solution should end up being the same if the regexp library is any good:
[21:37:48] <jgkamat> https://github.com/jgkamat/qutebrowser/commit/e37ecbdfddbeb142c5fa15ca93c3f1c07be8e503
[21:38:41] <jgkamat> (the regex package is still far from being fast enough to handle adblocking though, imo)
```
```
[18:08:19] <jgkamat> also, I spent a couple days looking at how ublock origin/abp do filter matching and I think their method might be an easier way to solve #4409
[18:11:08] <jgkamat> they essentially "tokenize" the URL with different methods, and each "token" maps to a list of regexes. That way, you only need to check whatever regexes match the current token. old ABP just used a 8 char substring, while new ABP splits the URL into parts (eg: https://qutebrowser.org/test -> https, qutebrowser, org, test). uBlock does the same thing as new abp except they create an int hash directly and they promote "hot" entires to
[18:11:08] <jgkamat> buckets more likely to match first
```

Tl;dr:

We can probably make regexp matching must faster, but `re` is pretty slow even in the optimal cases (and making people install libraries is annoying).

I'm guessing the best solution is what abp/ublock have implemented:

1. Take a URL: `http://www.example.com/test/index.html`
2. Tokenize it: `http`, `www`, `example`, `com`, `test`, `index`, `html`.
3. Loop through them (I'm guessing from most to least specific, ie: `html`->`index`->`test`->`example`->`com`->`www`->`http` and use the same algorithm that we currently have for tokens that hit
4. Run against 'slow rules' which aren't able to get any tokens.

The reason that I'm thinking this is better is because a single map lookup (especially failing ones) is usually much better than traversing a tree of maps. The critical part is making sure each bucket is mostly empty. Python handles strings in a weird way, so it might be actually better to use strings as keys (while ublock hashes the tokens to an int itself while tokenizing).

We won't have the luxury of returning early though, since we guarantee that the last entry matched will be returned (rather than does anything match at all).

Prior work (since ublock/abp source isn't very well documented):

-   [adblockparser](https://github.com/scrapinghub/adblockparser)
-   [adblock filter cheatsheet](https://adblockplus.org/filter-cheatsheet#options)
-   [ABP filter doc](https://adblockplus.org/en/filters)
-   [(ABP) Investigating filter matching algorithms](https://adblockplus.org/blog/investigating-filter-matching-algorithms)
-   [ABP faster filter matching](https://adblockplus.org/forum/viewtopic.php?t=6118)
-   [triegex](https://github.com/ZhukovAlexander/triegex)
-   [uBlock filtering code](https://github.com/gorhill/uBlock/blob/master/src/js/static-net-filtering.js)
-   [uBlock design doc](https://github.com/gorhill/uBlock/wiki/Overview-of-uBlock's-network-filtering-engine)
-   [ABP filtering code](https://github.com/adblockplus/adblockpluscore/blob/master/lib/matcher.js)
-   [ABP internal faq](https://adblockplus.org/faq_internal)
-   [Regexp performance tips](https://www.loggly.com/blog/five-invaluable-techniques-to-improve-regex-performance/)
-   [uBlock tokenizer](https://github.com/gorhill/uBlock/blob/261ef8c510fd91ead57948d1f7793a7a5e2a25fd/src/js/utils.js)

Of course, if we build a high performance library for this kind of thing (adblockparser dosen't look fast enough), we can probably share a lot of code between an adblocker and the url matcher.

**Comment by kmike** [comment]
> adblockparser dosen't look fast enough

@jgkamat regarding adblockparser - could you please share more details? were you using re2 library? it is supposed to be used with it, stdlib re is just a fallback.

**Comment by jgkamat** [comment]
Mikhail Korobov writes:

> @jgkamat regarding adblockparser - could you please share more details? were
> you using re2 library? it is supposed to be used with it, stdlib re is just
> a fallback.

I actually tried for quite a while, but I couldn't get the forked version of
pyre2 installed, with it giving me errors about missing variables in
'PyThreadState' on python 3.7, and on python 3.6, importing gave me
'NameError: basestring'.

Facebook's version installs and works fine, but leads to this error when used
in adblockparser (which I'm too lazy to fix). We would need the forked version
if we were to do this anyway, since we need to know which URL Pattern matched
(lastgroup/lastindex).

TypeError: compile() got an unexpected keyword argument 'flags'.

re2 is a fantastic regular expression engine, but I doubt it can handle
blindly merging together all the lists perfectly. For example, in the readme,
you say:

> Sometimes pyre2 prints something like re2/dfa.cc:459: DFA out of memory:
> prog size 270515 mem 1713850 to stderr. Give re2 library more memory to fix
> that:

while the lookup could be relatively fast (which depends on how advanced re2
is), the cost is moved into the optimizer step (both in memory and time),
which could be a problem depending on the size of the list (while that is
preferred generally, we don't want to hang while the dfa is built). In theory,
re2 could share space for the DFA between many parts (eg: all the scheme
matches), but I'm not sure if it's that smart to do that for duplicated regexp
groups. I'm wondering if we can coerce the regex engine into being smarter by
using something like the trigex project I linked above-though (which is
possible since we only allow generic glob wildcards), and if that's good
enough for our purposes.

Of course, this is all ignoring that if we 'can have' c extensions, we
wouldn't necessarily be doing it to get a better regex engine, but instead
some sort of rabin-karb style token matching (followed by infrequent regex
matches) which will be much faster (traversing an array plus some failing
integer hashtable lookups is much more cache-friendly than a DFA traversal,
which can miss on every character), which it looks like brave has largely done
for already (#4543):

https://github.com/brave/ad-block

However, for something that needs to go in core (URL patterns), I would much
prefer pure python, as adding a hard dependency on something like pyre2 or
brave's adblocker will complicate packaging greatly and may lead to many
distributions dropping our package (or just not updating it).

> If we disregard rule options, with re2 lookup speed is O(1) regarding a
> number of rules in easylist, while with regular re library it is O(N).

Because the majority (number) of requests are not blocked (or in this issue's
case, won't be matched for per-domain settings), optimizing for that case is
better than optimizing for the absolute worst case. Even if re2 is perfect, I
doubt a DFA traversal is faster than tokenizing a url and a couple hash table
misses. In many cases, we can further simplify many rules down to 'startswith'
or 'endswith' against the whole URL or hostname (which is what ublock does),
which must be faster than iterating over the entire URL, especially in the
failing case.

Do you have any latency numbers and/or profiles for adblockparser? Because I
couldn't get pyre2 working at all, I couldn't test it properly myself. uBlock
(claims to, with easylist) add on average ~0.1ms per request while adblock
plus adds on average ~0.4 ms, which seems to be close to what I can get with a
similar algorithm in pure-python with stdlib re (ranging from 0.01ms to ~2ms).
I'm pretty sure that brave's adblocker is much faster than any of these for
network filtering though.

**Comment by The-Compiler** [comment]
Reopening this for now - while we do have something similar like the `UrlPatternSet` in my original post, I still think there's more stuff needing some investigation (and maybe some low-hanging fruit).
